File: src/core/query.ts
import type { Database } from "bun:sqlite";
import type { Decision, DecisionType, DecisionStatus } from "./types";

interface DecisionRow {
  id: string;
  created_at: string;
  type: string;
  problem: string;
  choice: string;
  rationale: string | null;
  decided_by_role: string;
  decided_by_identifier: string | null;
  status: string;
  supersedes: string | null;
  superseded_by: string | null;
  raw_json: string;
}

function rowToDecision(row: DecisionRow): Decision {
  return JSON.parse(row.raw_json) as Decision;
}

export function queryById(db: Database, id: string): Decision | null {
  const row = db.query<DecisionRow, [string]>(
    "SELECT * FROM decisions WHERE id = ?"
  ).get(id);

  return row ? rowToDecision(row) : null;
}

export function queryByFile(db: Database, filePath: string): Decision[] {
  // Support glob patterns with LIKE
  const pattern = filePath.includes("*")
    ? filePath.replace(/\*/g, "%")
    : filePath;

  const rows = db.query<DecisionRow, [string]>(`
    SELECT d.* FROM decisions d
    INNER JOIN decision_files df ON d.id = df.decision_id
    WHERE df.file_path LIKE ?
    AND d.status = 'active'
    ORDER BY d.created_at DESC
  `).all(pattern);

  return rows.map(rowToDecision);
}

export function queryBySymbol(db: Database, symbol: string): Decision[] {
  const rows = db.query<DecisionRow, [string]>(`
    SELECT d.* FROM decisions d
    INNER JOIN decision_symbols ds ON d.id = ds.decision_id
    WHERE ds.symbol = ?
    AND d.status = 'active'
    ORDER BY d.created_at DESC
  `).all(symbol);

  return rows.map(rowToDecision);
}

export function queryByBead(db: Database, beadId: string): Decision[] {
  const rows = db.query<DecisionRow, [string]>(`
    SELECT d.* FROM decisions d
    INNER JOIN decision_beads db ON d.id = db.decision_id
    WHERE db.bead_id = ?
    AND d.status = 'active'
    ORDER BY d.created_at DESC
  `).all(beadId);

  return rows.map(rowToDecision);
}

export interface QueryOptions {
  type?: DecisionType;
  status?: DecisionStatus;
  limit?: number;
}

export function queryAll(db: Database, options: QueryOptions = {}): Decision[] {
  let sql = "SELECT * FROM decisions WHERE 1=1";
  const params: string[] = [];

  if (options.type) {
    sql += " AND type = ?";
    params.push(options.type);
  }

  if (options.status) {
    sql += " AND status = ?";
    params.push(options.status);
  }

  sql += " ORDER BY created_at DESC";

  if (options.limit) {
    sql += ` LIMIT ${options.limit}`;
  }

  const stmt = db.query<DecisionRow, string[]>(sql);
  const rows = stmt.all(...params);

  return rows.map(rowToDecision);
}

export function searchFullText(
  db: Database,
  query: string,
  options: QueryOptions = {}
): Decision[] {
  // Use FTS5 for full-text search
  let sql = `
    SELECT d.* FROM decisions d
    INNER JOIN decisions_fts fts ON d.id = fts.id
    WHERE decisions_fts MATCH ?
  `;
  const params: string[] = [query];

  if (options.type) {
    sql += " AND d.type = ?";
    params.push(options.type);
  }

  if (options.status) {
    sql += " AND d.status = ?";
    params.push(options.status);
  }

  sql += " ORDER BY rank";

  if (options.limit) {
    sql += ` LIMIT ${options.limit}`;
  }

  try {
    const stmt = db.query<DecisionRow, string[]>(sql);
    const rows = stmt.all(...params);
    return rows.map(rowToDecision);
  } catch (e) {
    // FTS query might fail on invalid syntax, fall back to LIKE search
    return searchLike(db, query, options);
  }
}

function searchLike(
  db: Database,
  query: string,
  options: QueryOptions = {}
): Decision[] {
  const pattern = `%${query}%`;

  let sql = `
    SELECT * FROM decisions
    WHERE (problem LIKE ? OR choice LIKE ? OR rationale LIKE ?)
  `;
  const params: string[] = [pattern, pattern, pattern];

  if (options.type) {
    sql += " AND type = ?";
    params.push(options.type);
  }

  if (options.status) {
    sql += " AND status = ?";
    params.push(options.status);
  }

  sql += " ORDER BY created_at DESC";

  if (options.limit) {
    sql += ` LIMIT ${options.limit}`;
  }

  const stmt = db.query<DecisionRow, string[]>(sql);
  const rows = stmt.all(...params);

  return rows.map(rowToDecision);
}

export function getActiveConstraints(db: Database): Decision[] {
  const rows = db.query<DecisionRow, []>(`
    SELECT * FROM decisions
    WHERE type = 'constraint' AND status = 'active'
    ORDER BY created_at DESC
  `).all();

  return rows.map(rowToDecision);
}

export function getDecisionsForContext(
  db: Database,
  path: string
): { decisions: Decision[]; constraints: Decision[] } {
  const decisions = queryByFile(db, path);
  const constraints = getActiveConstraints(db);

  return { decisions, constraints };
}

File: internal/query/query.go
package query

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/tyroneavnit/keel/internal/index"
	"github.com/tyroneavnit/keel/internal/types"
)

// Options for querying decisions
type Options struct {
	Type   string
	Status string
	Limit  int
}

// ContextResult contains decisions and constraints for a given context
type ContextResult struct {
	Decisions   []*types.Decision
	Constraints []*types.Decision
}

func rowToDecision(rawJSON string) (*types.Decision, error) {
	var d types.Decision
	if err := json.Unmarshal([]byte(rawJSON), &d); err != nil {
		return nil, err
	}
	return &d, nil
}

// ByID queries a decision by its ID
func ByID(db *index.DB, id string) (*types.Decision, error) {
	var rawJSON string
	err := db.QueryRow("SELECT raw_json FROM decisions WHERE id = ?", id).Scan(&rawJSON)
	if err != nil {
		return nil, nil // Not found
	}
	return rowToDecision(rawJSON)
}

// ByFile queries decisions affecting a file path
func ByFile(db *index.DB, filePath string) ([]*types.Decision, error) {
	// Support glob patterns with LIKE
	pattern := filePath
	if strings.Contains(filePath, "*") {
		pattern = strings.ReplaceAll(filePath, "*", "%")
	}

	rows, err := db.Query(`
		SELECT d.raw_json FROM decisions d
		INNER JOIN decision_files df ON d.id = df.decision_id
		WHERE df.file_path LIKE ?
		AND d.status = 'active'
		ORDER BY d.created_at DESC
	`, pattern)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var decisions []*types.Decision
	for rows.Next() {
		var rawJSON string
		if err := rows.Scan(&rawJSON); err != nil {
			continue
		}
		if d, err := rowToDecision(rawJSON); err == nil {
			decisions = append(decisions, d)
		}
	}

	return decisions, nil
}

// BySymbol queries decisions for a symbol
func BySymbol(db *index.DB, symbol string) ([]*types.Decision, error) {
	rows, err := db.Query(`
		SELECT d.raw_json FROM decisions d
		INNER JOIN decision_symbols ds ON d.id = ds.decision_id
		WHERE ds.symbol = ?
		AND d.status = 'active'
		ORDER BY d.created_at DESC
	`, symbol)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var decisions []*types.Decision
	for rows.Next() {
		var rawJSON string
		if err := rows.Scan(&rawJSON); err != nil {
			continue
		}
		if d, err := rowToDecision(rawJSON); err == nil {
			decisions = append(decisions, d)
		}
	}

	return decisions, nil
}

// ByRef queries decisions linked to a reference ID
func ByRef(db *index.DB, refID string) ([]*types.Decision, error) {
	rows, err := db.Query(`
		SELECT d.raw_json FROM decisions d
		INNER JOIN decision_refs dr ON d.id = dr.decision_id
		WHERE dr.ref_id = ?
		AND d.status = 'active'
		ORDER BY d.created_at DESC
	`, refID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var decisions []*types.Decision
	for rows.Next() {
		var rawJSON string
		if err := rows.Scan(&rawJSON); err != nil {
			continue
		}
		if d, err := rowToDecision(rawJSON); err == nil {
			decisions = append(decisions, d)
		}
	}

	return decisions, nil
}

// All queries all decisions with optional filters
func All(db *index.DB, opts Options) ([]*types.Decision, error) {
	sql := "SELECT raw_json FROM decisions WHERE 1=1"
	var args []interface{}

	if opts.Type != "" {
		sql += " AND type = ?"
		args = append(args, opts.Type)
	}

	if opts.Status != "" {
		sql += " AND status = ?"
		args = append(args, opts.Status)
	}

	sql += " ORDER BY created_at DESC"

	if opts.Limit > 0 {
		sql += fmt.Sprintf(" LIMIT %d", opts.Limit)
	}

	rows, err := db.Query(sql, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var decisions []*types.Decision
	for rows.Next() {
		var rawJSON string
		if err := rows.Scan(&rawJSON); err != nil {
			continue
		}
		if d, err := rowToDecision(rawJSON); err == nil {
			decisions = append(decisions, d)
		}
	}

	return decisions, nil
}


// ActiveConstraints returns all active constraint decisions
func ActiveConstraints(db *index.DB) ([]*types.Decision, error) {
	rows, err := db.Query(`
		SELECT raw_json FROM decisions
		WHERE type = 'constraint' AND status = 'active'
		ORDER BY created_at DESC
	`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var decisions []*types.Decision
	for rows.Next() {
		var rawJSON string
		if err := rows.Scan(&rawJSON); err != nil {
			continue
		}
		if d, err := rowToDecision(rawJSON); err == nil {
			decisions = append(decisions, d)
		}
	}

	return decisions, nil
}

// ForContext returns decisions and constraints for a given file path
func ForContext(db *index.DB, path string) (*ContextResult, error) {
	decisions, err := ByFile(db, path)
	if err != nil {
		return nil, err
	}

	constraints, err := ActiveConstraints(db)
	if err != nil {
		return nil, err
	}

	return &ContextResult{
		Decisions:   decisions,
		Constraints: constraints,
	}, nil
}

// RefLink represents a decision-to-ref relationship
type RefLink struct {
	DecisionID string
	RefID      string
}

// AllRefs returns all decision-to-ref links
func AllRefs(db *index.DB) ([]RefLink, error) {
	rows, err := db.Query(`SELECT decision_id, ref_id FROM decision_refs`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var links []RefLink
	for rows.Next() {
		var link RefLink
		if err := rows.Scan(&link.DecisionID, &link.RefID); err != nil {
			continue
		}
		links = append(links, link)
	}
	return links, nil
}

// FileLink represents a decision-to-file relationship
type FileLink struct {
	DecisionID string
	FilePath   string
}

// AllFileLinks returns all decision-to-file links
func AllFileLinks(db *index.DB) ([]FileLink, error) {
	rows, err := db.Query(`SELECT decision_id, file_path FROM decision_files`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var links []FileLink
	for rows.Next() {
		var link FileLink
		if err := rows.Scan(&link.DecisionID, &link.FilePath); err != nil {
			continue
		}
		links = append(links, link)
	}
	return links, nil
}
